weaviate/embedded.py

```diff
@@ -0,0 +1,126 @@
+import hashlib
+import subprocess
+import os
+import signal
+import stat
+import time
+from urllib.parse import urlparse, parse_qsl
+import urllib.request
+from pathlib import Path
+import socket
+
+from weaviate.exceptions import WeaviateStartUpError
+
+weaviate_binary_path = "./weaviate-server-embedded"
+weaviate_binary_url = (
+    "https://github.com/samos123/weaviate/releases/download/v1.17.3/weaviate-server"
+)
+weaviate_persistence_data_path = "./weaviate-data"
+weaviate_binary_md5 = "38b8ac3c77cc8707999569ae3fe34c71"
+
+
+class EmbeddedDB:
+    # TODO add a stop function that gets called when python process exits
+    def __init__(self, url: str = ""):
+        parsed = parse_qsl(urlparse(url).query)
+        parsed = dict(parsed)
+        self.port = int(parsed.get("port", 6666))
+        self.pid = 0
+
+    def __del__(self):
+        self.stop()
+
+    def ensure_weaviate_binary_exists(self):
+        file = Path(weaviate_binary_path)
+        if not file.exists():
+            print(
+                f"Binary {weaviate_binary_path} did not exist. "
+                f"Downloading binary from {weaviate_binary_url}"
+            )
+            urllib.request.urlretrieve(weaviate_binary_url, weaviate_binary_path)
+            # Ensuring weaviate binary is executable
+            file.chmod(file.stat().st_mode | stat.S_IEXEC)
+            with open(file, "rb") as f:
+                assert (
+                    hashlib.md5(f.read()).hexdigest() == weaviate_binary_md5
+                ), f"md5 of binary {weaviate_binary_path} did not match {weaviate_binary_md5}"
+
+    def is_running(self) -> bool:
+        binary_name = os.path.basename(weaviate_binary_path)
+        result = subprocess.run(
+            f"ps aux | grep {binary_name} | grep -v grep", shell=True, capture_output=True
+        )
+        if result.returncode == 1:
+            return False
+        else:
+            output = result.stdout.decode("ascii")
+            if binary_name in output:
+                return True
+        return False
+
+    def is_listening(self) -> bool:
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            s.connect(("127.0.0.1", self.port))
+            s.close()
+            return True
+        except socket.error:
+            s.close()
+            return False
+
+    def wait_till_listening(self):
```

Comment:
hmm but I don't have access to the connection object

Reply:
I tried this in a separate branch but it made the unit tests of test_embedded.py very hard since connection class will have many requests during initialization that would fail, see my other branch here: embedded-reuseconnection

So I suggest to keep this part as is and just use the wait_for_listening that I built. You good with that?

Reply:
If it is just the test, would a mock work? If not let's leave it like it is

Reply:
How do I test if weaviate came up with if I mock the client/connection? I need an actual client or connection to check if Weaviate came up or not. You can see in my branch I was struggling ha, you're right though it was mostly the testing piece that got tricky. My manual tests showed it did work as expected. There was a circular dependency due to type hints as well but got that resolved.

I would prefer to leave it like it is where embedded has it's own check to see if server is listening. Also open to do a pair programming session on this.

Reply:
let's deal with that another time, we can always improve/remove that function later